pkg/controller/rbac/application/managed.go

```diff
@@ -0,0 +1,195 @@
+)
+
+// Error strings.
+const (
+	errNotApplication    = "managed resource is not an Application"
+	errCreateApplication = "cannot create Application"
+	errUpdateApplication = "cannot update Application"
+	errGetApplication    = "cannot get Application"
+	errDeleteApplication = "cannot delete Application"
+)
+
+// Setup adds a controller that reconciles Application.
+func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.ApplicationKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.Application{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.ApplicationGroupVersionKind),
+			// Override default initializers in case to remove NewNameAsExternalName Initializer
+			managed.WithInitializers(),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	creds, _, err := azureclients.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+	cfg, err := adal.NewOAuthConfig(creds[azure.CredentialsKeyActiveDirectoryEndpointURL], creds[azure.CredentialsKeyTenantID])
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create OAuth configuration")
+	}
+	token, err := adal.NewServicePrincipalToken(*cfg,
+		creds[azure.CredentialsKeyClientID],
+		creds[azure.CredentialsKeyClientSecret],
+		creds[azure.CredentialsKeyActiveDirectoryGraphResourceID])
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create service principal token")
+	}
+	if err := token.Refresh(); err != nil {
+		return nil, errors.Wrap(err, "cannot refresh service principal token")
+	}
+	ta := autorest.NewBearerAuthorizer(token)
+	ac := graphrbac.NewApplicationsClient(creds[azure.CredentialsKeyTenantID])
+	ac.Authorizer = ta
+	return &external{c: ac}, nil
+}
+
+type external struct {
+	c graphrbacapi.ApplicationsClientAPI
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	s, ok := mg.(*v1alpha1.Application)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotApplication)
+	}
+	if meta.GetExternalName(s) == "" {
+		return managed.ExternalObservation{}, nil
+	}
+
+	az, err := e.c.Get(ctx, meta.GetExternalName(s))
+	if err != nil {
+		err = errors.Wrap(err, errGetApplication)
+		err = resource.Ignore(azureclients.IsNotFound, err)
```

Comment:
Looks like this will never return true. `azure.IsNotFound` does a type assertion for `autorest.DetailedError` on the error passed as its argument. But we will have an `errors.withStack` (a wrapped error) and never have an `autorest.DetailedError`. I would suggest the usual processing done elsewhere here. As an example please refer to:
https://github.com/crossplane/provider-azure/blob/1c0702110ed45feabe2af15a0478e753d1c34d4e/pkg/controller/network/virtualnetwork/managed.go#L98